File: examples/04-java-threads/Example3/Example3.java
// =================================================================
//
// File: Example3.java
// Author: Pedro Perez
// Description: This file contains the code to perform the numerical
//				integration of a function within a defined interval
//				using Threads.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

class Sin implements Function {
	public double eval(double x) {
		return Math.sin(x);
	}
}

public class Example3 extends Thread {
	private static final int RECTS = 1_000_000_000;
	private double x, dx, result;
	private int start, end;
	private Function fn;

	public Example3(double a, double b, Function fn, int start, int end) {
		this.x = Math.min(a, b);
		this.dx = (Math.max(a,b) - Math.min(a, b)) / RECTS;
		this.fn = fn;
		this.start = start;
		this.end = end;
	}

	public double getResult() {
		return result;
	}

	public void run() {
		result = 0;
		for (int i = start; i < end; i++) {
			result += fn.eval(x + (i * dx));
		}
		result = result * dx;
	}

	public static void main(String args[]) {
		long startTime, stopTime;
		int block;
		Example3 threads[];
		double ms, result = 0;

		block = RECTS / Utils.MAXTHREADS;
		threads = new Example3[Utils.MAXTHREADS];

		System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
		ms = 0;
		for (int j = 1; j <= Utils.N; j++) {
			for (int i = 0; i < threads.length; i++) {
				if (i != threads.length - 1) {
					threads[i] = new Example3(0, Math.PI, new Sin(), (i * block), ((i + 1) * block));
				} else {
					threads[i] = new Example3(0, Math.PI, new Sin(), (i * block), RECTS);
				}
			}

			startTime = System.currentTimeMillis();
			for (int i = 0; i < threads.length; i++) {
				threads[i].start();
			}
			for (int i = 0; i < threads.length; i++) {
				try {
					threads[i].join();
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
			stopTime = System.currentTimeMillis();
			ms +=  (stopTime - startTime);

			if (j == Utils.N) {
				result = 0;
				for (int i = 0; i < threads.length; i++) {
					result += threads[i].getResult();
				}
			}
		}
		System.out.printf("result = %.5f\n", result);
		System.out.printf("avg time = %.5f\n", (ms / Utils.N));
	}
}

File: examples/07-tbb/example3.cpp
// =================================================================
//
// File: Example3.cpp
// Author: Pedro Perez
// Description: This file contains the code to perform the numerical
//				integration of a function within a defined interval
//				using Intel's TBB. To compile:
//				g++ example3.cpp -ltbb
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cmath>
#include <tbb/parallel_reduce.h>
#include <tbb/blocked_range.h>
#include "utils.h"

const double PI = 3.14159265;
const int RECTS = 1000000000; //1e9

using namespace std;
using namespace tbb;

double function(double x) {
	return sin(x);
}

class Integration {
private:
	double x, dx, result;
	double (*func) (double);

public:
	Integration(double xx, double dxx, double (*fn) (double))
		: x(xx), dx(dxx), func(fn), result(0) {}

	Integration(Integration &obj, split)
		: x(obj.x), dx(obj.dx), func(obj.func), result(0) {}

	double getResult() const {
		return result * dx;
	}

	// 	void operator() (const blocked_range<int> &r) const {
	void operator() (const blocked_range<int> &r) {
		for (int i = r.begin(); i != r.end(); i++) {
			result += func(x + (i * dx));
		}
	}

	void join(const Integration &x) {
		result += x.result;
	}
};

int main(int argc, char* argv[]) {
	double ms;
	double x, dx, result;

	x = 0;
	dx = (PI - 0.0) / RECTS;

	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		start_timer();

		Integration obj(x, dx, function);
		parallel_reduce(blocked_range<int>(0, RECTS), obj);
		result = obj.getResult();

		ms += stop_timer();
	}
	cout << "result = " << setprecision(15) << result << endl;
	cout << "avg time = " << setprecision(15) << (ms / N) << " ms" << endl;

	return 0;
}
